cgmanifest.json

```diff
@@ -0,0 +1,49 @@
+{
+    "$schema": "https://json.schemastore.org/component-detection-manifest.json",
```

Comment:
According to these schema, the `registrations`, `component` and `type` properties should all be lowercase. It's a little confusing because the definition names use Title casing, but the properties actuall use lowercase.

Similarly, `developmentDependency' should be in camelCase.

Reply:
I copied the template from our runtime file which is already tested. https://github.com/dotnet/runtime/blob/main/src/native/external/cgmanifest.json. should I follow the schema?

Reply:
Actually in the [schema](https://[json.schemastore.org/component-detection-manifest.json](https://json.schemastore.org/component-detection-manifest.json)), I am seeing it is using:

```json

        "Registration": {
        
        "Component": {        

```

Reply:
That's where I think it's misleading and confusing, but I'm pretty sure we need to switch to lowercase `registrations` and `component`. Tagging @eiriktsarpalis to chime in and confirm my understanding of this JSON Schema, but here's how I and reading the schema:

1. The schema is defining 'Definitions' of objects that can be contained within the manifest
    - Those Definitions are given human-readable names, e.g. "Component" and "Registration"
2. The "CGManifest" is defined to be an `object` with properties of `registrations` and `version`, both of which are required
    - The `registrations` property is defined to be an array of "Registration" definitions, where the definition of a "Registration" is also within this document
3. A "Registration" is defined to be an `object` with properties of `component`, `dependencyRoots`, and `developmentDependency`
    - Only `component` is required and it's defined as a "Component" definition, where it is also defined within this document
5. A "Component" definition is an `object` that needs to be one of the various component types, each of those options referring to another Definition within this document
6. The "Other" definition is an `object` with two properties: `type` (which can only be `"other"`), and then an `other` property as well
    - The `other` property is defined as an `object` with properties of `name`, `version`, `downloadUrl`, and `hash`, all of which are required

The human-readable friendly names of the definitions is where it's misleading. To check my work, I spun up a notebook that uses JsonSchema.Net to validate the document against the schema and it shows it's currently invalid. Then I updated the JSON and the revised content shows as being valid against the schema.

``` C#
using System;
using System.Net.Http;
using System.Text.Json.Nodes;
using System.Threading.Tasks;
using Json.Schema;

HttpClient httpClient = new();

string schemaUrl = "https://json.schemastore.org/component-detection-manifest.json";
HttpResponseMessage schemaResponse = await httpClient.GetAsync(schemaUrl);
schemaResponse.EnsureSuccessStatusCode();
string schemaJson = (await schemaResponse.Content.ReadAsStringAsync()).Replace("/draft-04/", "/draft-06/");
JsonSchema schema = JsonSchema.FromText(schemaJson);

string manifestUrl = "https://raw.githubusercontent.com/dotnet/machinelearning/88b6ce3565a81dba707029dc17e38297d6c9075a/cgmanifest.json";
HttpResponseMessage manifestResponse = await httpClient.GetAsync(manifestUrl);
manifestResponse.EnsureSuccessStatusCode();
string manifestJson = await manifestResponse.Content.ReadAsStringAsync();
JsonNode manifest = JsonNode.Parse(manifestJson);

var result = schema.Evaluate(manifest);
Console.WriteLine($"Valid: {result.IsValid}");
// Valid: False

string lowercaseManifestJson = manifestJson
    .Replace("\"Registrations\"", "\"registrations\"")
    .Replace("\"Component\"", "\"component\"");

JsonNode lowercaseManifest = JsonNode.Parse($$"""
{
    "$schema": "https://json.schemastore.org/component-detection-manifest.json",
    "version": 1,
    "registrations": [
      {
        "component": {
          "type": "other",
          "other": {
            "name": "cl100k_base.tiktoken",
            "downloadUrl": "https://openaipublic.blob.core.windows.net/encodings/cl100k_base.tiktoken",
            "hash": "sha1:6494e42d5aad2bbb441ea9793af9e7db335c8d9c",
            "version": "version-needed"
          }
        },
        "developmentDependency": false
      },
      {
        "component": {
          "type": "other",
          "other": {
            "name": "o200k_base.tiktoken",
            "downloadUrl": "https://openaipublic.blob.core.windows.net/encodings/o200k_base.tiktoken",
            "hash": "sha1:1d4fdeb17c52829ead47ac65e61197fd530b1c31",
            "version": "version-needed"
          }
        },
        "developmentDependency": false
      },
      {
        "component": {
          "type": "other",
          "other": {
            "name": "p50k_base.tiktoken",
            "downloadUrl": "https://openaipublic.blob.core.windows.net/encodings/p50k_base.tiktoken",
            "hash": "sha1:0ecf4ae6d454e7719bcf35f284eac0b73f37e3c9",
            "version": "version-needed"
          }
        },
        "developmentDependency": false
      },
      {
        "component": {
          "type": "other",
          "other": {
            "name": "r50k_base.tiktoken",
            "downloadUrl": "https://openaipublic.blob.core.windows.net/encodings/r50k_base.tiktoken",
            "hash": "sha1:5674ba48e48e76284eb747c896a291dc5583c808",
            "version": "version-needed"
          }
        },
        "developmentDependency": false
      }
    ]
}
""");

var lowercaseResult = schema.Evaluate(lowercaseManifest);
Console.WriteLine($"Valid: {lowercaseResult.IsValid}");
// Valid: True
```

These were the changes needed to make it valid:

1. Change `"Registrations"` to `"registrations"`
2. Change each `"Component"` to `"component"`
3. Add a top-level `version` property to the document (I'm not sure what value we should use for this)
4. Add a `version` property to each of the `component` objects (I'm not sure what values we should use)

Reply:
@jeffhandley is correct. Identifiers under the `"definitions"` keyword are implementation detail of the particular schema document and should not have impact on the actual schema being specified.

Reply:
Thanks @jeffhandley. Looks like we already have issues with the file used in the runtime repo then. Or the tool currently reading this file is not really checking against the schema?

Reply:
I have addressed all feedback and ran the schema validation code too.